docs: Consistent type names in RPC help descriptions

ch4ot1c opened the pull request:
Followup to https://github.com/bitcoin/bitcoin/pull/14373.

Now, only these appear in description text:
```
boolean
numeric
string

json array
json array of xs
json object
json object of xs
```

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -684,12 +684,12 @@ static UniValue getbalance(const JSONRPCRequest& request)
     if (request.fHelp || (request.params.size() > 3 ))
         throw std::runtime_error(
-            "getbalance ( \"(dummy)\" minconf include_watchonly )\n"
+            "getbalance ( dummy minconf include_watchonly )\n"
```
`dummy` is a string.

ch4ot1c commented:
#### Final notes:

(numeric) or (int)
(boolean)
(string)

(x, optional)
(x, optional, default=y)
(x, optional, default=\"y\")

(x, required [for ...])

(array)
(array of x[s])
(object)

Only these appear as types now, in both in requests and results.

Additional notes:
- Any usage of an ` = ` within RPC descriptions has been trimmed
- Suffixed `Default: ` cases have been moved to the type string e.g. `(x, optional, default=y)` in RPCs
- The json object for label & purpose data is now named 'Address information' in both `getaddressinfo` and `getaddressesbylabel`

- `Must be one of:\n prefix should be present for all argument option lists >= 3
- `Results` still has different/inconsistent formatting throughout, which will need to be addressed separately.

MarcoFalke commented:
* importprunedfunds has arguments, but they are not listed in the help string. Mind adding them?
* uptime takes (and ignores) a dummy argument, whereas it shouldn't. Mind fixing the code to throw when a dummy is provided?

MarcoFalke commented on src/rpc/server.cpp:
```diff
@@ -245,6 +245,11 @@ static UniValue uptime(const JSONRPCRequest& jsonRequest)
+    if (!dummy_value.isNull()) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "No dummy first argument may be included.");
+    }
+
```
Can be fixed instead by replacing `> 1` with `> 0` a few lines up?

ch4ot1c replied:
Ah, that was what I thought at first.

MarcoFalke commented:
`prioritisetransaction` has an argument with a space. Mind to replace the space with underscore?

MarcoFalke commented on src/wallet/rpcdump.cpp:
```diff
@@ -1108,8 +1113,8 @@ UniValue importmulti(const JSONRPCRequest& mainRequest)
             "importmulti \"requests\" ( \"options\" )\n\n"
```
```suggestion
             "importmulti \"requests\" ( \"options\" )\n"
```

ch4ot1c commented:
`prioritisetransaction` and `getbalance` changes are already in the PR.

MarcoFalke commented on src/zmq/zmqrpc.cpp:
```diff
@@ -20,7 +20,7 @@ UniValue getzmqnotifications(const JSONRPCRequest& request)
             "\nResult:\n"
             "[\n"
-            "  {                        (json object)\n"
+            "  {                        (object)\n"
```
Half of the changes seem to be of this fashion. Mind to create a primitive scripted-diff commit in a separate pull request to get the easy to review things out of the way?

ch4ot1c replied:
Damn, they need to make a `sed`-builder plugin alongside find-and-replace in vscode, that supports negation. Way easier workflow for newer contributors than all the nuanced escape characters that are necessary in *nix tools.

Is it preferred to use `git grep (-l) (-L)` or `find . x` in scripted diffs?

Also, does this really make it easier to review for correctness? I'm not so sure.

Next commit will be a scripted diff for:
> Replace `(boolean)W` with `(boolean) W` in `.`

`git grep -l "(boolean)W" | xargs sed -i "s/(boolean)W/(boolean) W/g"`
is about right I think?

What if I'm stuck in PowerShell! No `xargs` available.

ch4ot1c replied:
And on Mac, `gsed` is needed to replicate `sed` behavior performed in scripted-diffs

MarcoFalke replied:
I'd do it like this: `sed -i --regexp-extended -e 's/\(json (array|object)\)/(\1)/g' $(git grep -l '(json ')`. There might be a hack to make it work on macOS, but we mostly care that it runs on Ubuntu et al.

ch4ot1c replied:
Pushed to https://github.com/ch4ot1c/bitcoin/tree/fix/rpc-descriptions, but the commit isn't appearing here. Strange...

ch4ot1c commented:
Pushed to https://github.com/ch4ot1c/bitcoin/tree/fix/rpc-descriptions, but the commit isn't appearing here. Strange...

MarcoFalke commented:
Could rebase and squash with something like:

```sh
git fetch bitcoin
git checkout fix/rpc-descriptions
git merge bitcoin/master
git reset --soft bitcoin/master
git commit -m "More RPC help description fixes"
git push origin fix/rpc-descriptions --force

laanwj commented:
utACK 2f6fff743a2f1e7c51bee41c18498b4043ba9e48

MarcoFalke commented:
`src/rpc/util.cpp` uses "json object". Imo everything should use the same

ch4ot1c commented:
@MarcoFalke I see. To confirm, `json array` and `json object` everywhere applicable, preferred over `array` and `object`, aka `util.cpp` is set in stone? Will redo as a scripted-diff once more.

ch4ot1c committed:
scripted-diff: [rpc] Descriptions: 'array of' -> 'json array of'

-BEGIN VERIFY SCRIPT-

git grep -l "(array of" src | xargs sed -i "s/(array of/(json array of/g"

-END VERIFY SCRIPT-

ch4ot1c committed:
scripted-diff: [rpc] Descriptions: 'of Objects' -> 'of json objects'

-BEGIN VERIFY SCRIPT-

git grep -l "of Objects" src | xargs sed -i "s/of Objects/of json objects/g"

-END VERIFY SCRIPT-

ch4ot1c committed:
scripted-diff: [rpc] Descriptions: Fix usage of 'list' instead of 'array'

-BEGIN VERIFY SCRIPT-

git grep -l "(list of objects)" src | xargs sed -i "s/(list of objects)/(json array of json objects)/g"

-END VERIFY SCRIPT-

ch4ot1c committed:
scripted-diff: [rpc] Descriptions: '(bool) ' -> '(boolean) '

-BEGIN VERIFY SCRIPT-

git grep -l "(bool) " src | xargs sed -i "s/(bool) /(boolean) /g"

-END VERIFY SCRIPT-

ch4ot1c committed:
scripted-diff: [rpc] Descriptions: 'of objects' -> 'of json objects'

-BEGIN VERIFY SCRIPT-

git grep -l "of objects[)\.]" src/rpc src/wallet | xargs sed -i "s/of objects/of json objects/g"

-END VERIFY SCRIPT-

ch4ot1c committed:
scripted-diff: [rpc] Descriptions: ' an object' -> ' a json object'

-BEGIN VERIFY SCRIPT-

git grep -l " an object" src/rpc src/wallet | xargs sed -i "s/ an object/ a json object/g"

-END VERIFY SCRIPT-

ch4ot1c committed:
scripted-diff: [rpc] Descriptions: 'of string)' -> 'of strings)'

-BEGIN VERIFY SCRIPT-

git grep -l "of string)" src/rpc src/wallet | xargs sed -i "s/of string)/of strings)/g"

-END VERIFY SCRIPT-

ch4ot1c committed:
scripted-diff: [rpc] Descriptions: 'A list of ' -> 'An array of ' in mining.cpp

-BEGIN VERIFY SCRIPT-

git grep -l "A list of " src/rpc/mining.cpp | xargs sed -i "s/A list of /An array of /g"

-END VERIFY SCRIPT-

ch4ot1c committed:
[rpc] Descriptions: (int) -> (numeric) in blockchain.cpp

ch4ot1c committed:
[rpc] Descriptions: 'integer(s)' -> 'numeric(s)'

ch4ot1c committed:
[rpc] Descriptions: Use 'json object' instead of 'dict' in util.h

ch4ot1c commented on src/rpc/mining.cpp:
```diff
@@ -314,12 +314,12 @@ static UniValue getblocktemplate(const JSONRPCRequest& request)
                         {
                             {"mode", RPCArg::Type::STR, /* opt */ true, /* default_val */ "", "This must be set to \"template\", \"proposal\" (see BIP 23), or omitted"},
-                            {"capabilities", RPCArg::Type::ARR, /* opt */ true, /* default_val */ "", "A list of strings",
+                            {"capabilities", RPCArg::Type::ARR, /* opt */ true, /* default_val */ "", "An array of strings",
```
`A json array` here instead?

ch4ot1c committed:
scripted-diff: [rpc] Descriptions: 'Must be one of' -> 'Must be one of:'

-BEGIN VERIFY SCRIPT-

git grep -l "Must be one of\\\n" src | xargs sed -i "s/Must be one of\\\n/Must be one of:\\\n/g"

-END VERIFY SCRIPT-

ch4ot1c committed:
scripted-diff: [rpc] Descriptions: 'an array' -> 'a json array'

-BEGIN VERIFY SCRIPT-

git grep -l "[aA]n array" src/rpc src/wallet | xargs sed -i "s/an array/a json array/g"

git grep -l "[aA]n array" src/rpc src/wallet | xargs sed -i "s/An array/A json array/g"

-END VERIFY SCRIPT-

ch4ot1c committed:
scripted-diff: [rpc] Tests: 'pair not an object' -> 'pair not a json object'

-BEGIN VERIFY SCRIPT-

git grep -l "pair not an object" test/functional/rpc_rawtransaction.py | xargs sed -i "s/pair not an object/pair not a json object/g"

-END VERIFY SCRIPT-

ch4ot1c committed:
scripted-diff: [rpc] Descriptions: 'of numeric' -> 'of numerics' in blockchain.cpp

-BEGIN VERIFY SCRIPT-

git grep -l "of numeric" src/rpc/blockchain.cpp | xargs sed -i "s/of numeric/of numerics/g"

-END VERIFY SCRIPT-

ch4ot1c committed:
scripted-diff: [rpc] Descriptions: '(object) ' -> '(json object) ' in src/rpc, src/wallet

-BEGIN VERIFY SCRIPT-

git grep -l "(object) " src/wallet src/rpc | xargs sed -i "s/(object) /(json object) /g"

-END VERIFY SCRIPT-

laanwj commented:
Concept ACK.

Though I think the jargon switch to `numeric` loses information.

The code, in practice, makes a difference between `numeric` (in general) and `integer`. For example where `get_int` is used, only an integer will be accepted. Not a decimal value with a point.
